Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp

```diff
@@ -314,6 +320,75 @@ MediaStreamTrack::TrackCapabilities MediaStreamTrack::getCapabilities() const
     return result;
 }
 
+void MediaStreamTrack::queueAndProcessSerialAction(Function<Ref<GenericPromise>(SerialActionResolution)>&& action)
+{
+    ASSERT(isMainThread());
+    m_pendingActionQueue.append(WTFMove(action));
+    processNextSerialAction();
+}
+
+void MediaStreamTrack::processNextSerialAction()
+{
+    ASSERT(isMainThread());
+    if (m_currentSerialAction || m_pendingActionQueue.isEmpty())
+        return;
+
+    m_currentSerialAction = { m_pendingActionQueue.takeFirst() };
+    m_currentSerialAction.value()(SerialActionResolution::Run)->whenSettled(RunLoop::main(), [protectedThis = Ref { *this }] {
+        protectedThis->m_currentSerialAction = { };
+        protectedThis->processNextSerialAction();
+    });
+}
+
+auto MediaStreamTrack::takePhoto(PhotoSettings&& settings) -> Ref<TakePhotoPromise>
+{
+    TakePhotoPromise::Producer producer;
+    Ref<TakePhotoPromise> promise = producer;
+
+    queueAndProcessSerialAction([settings = WTFMove(settings), protectedThis = Ref { *this }, producer = WTFMove(producer)](SerialActionResolution resolution) mutable -> Ref<GenericPromise> {
+
+        switch (resolution) {
+        case SerialActionResolution::Run:
+
+            // https://w3c.github.io/mediacapture-image/#dom-imagecapture-takephoto
+            // If the readyState of track provided in the constructor is not live, return
+            // a promise rejected with a new DOMException whose name is InvalidStateError,
+            // and abort these steps.
+
+            if (protectedThis->m_readyState != State::Live) {
+                producer.reject(Exception { InvalidStateError, "Track has ended"_s });
+                return GenericPromise::createAndResolve();
+            }
+
+            return protectedThis->m_private->takePhoto(WTFMove(settings))->whenSettled(RunLoop::main(),
+                [protectedThis = WTFMove(protectedThis), producer = WTFMove(producer)] (auto&& result) mutable {
+
+                // https://w3c.github.io/mediacapture-image/#dom-imagecapture-takephoto
+                // If the operation cannot be completed for any reason (for example, upon
+                // invocation of multiple takePhoto() method calls in rapid succession),
+                // then reject p with a new DOMException whose name is UnknownError, and
+                // abort these steps.
+                if (!result)
+                    producer.reject(Exception { UnknownError, WTFMove(result.error()) });
+                else if (RefPtr context = protectedThis->scriptExecutionContext(); !context || context->activeDOMObjectsAreStopped() || protectedThis->m_ended)
+                    producer.reject(Exception { OperationError, "Track has ended"_s });
+                else
+                    producer.resolve(WTFMove(result.value()));
```

Comment:
I would tend to create the blob here instead of the call site.

Reply:
I suggested creating the blob at the point where the DOMPromise is resolved, as it clearly separate the JS side of things from the implementation.

Also allows to avoid to unnecessarily link all JS headers that you don't have to use.

To me, resolving the DOM promise at the same level the DOMPromise was first received makes for clearer reading/review.